domainbed/algorithms.py

```diff
@@ -119,6 +120,141 @@ def update(self, minibatches, unlabeled=None):
     def predict(self, x):
         return self.network(x)
 
+class MAT(Algorithm):
+
+    def __init__(self, input_shape, num_classes, num_domains, hparams, train_loaders):
+        super(MAT, self).__init__(input_shape, num_classes, num_domains, hparams)
+
+        # Check if the train loader is provided
+        assert train_loaders is not None, "Train loaders must be provided for MAT"
+
+        self.featurizer = networks.Featurizer(input_shape, self.hparams)
+        self.classifier_c = networks.Classifier(self.featurizer.n_outputs, num_classes, 
+                                                self.hparams['nonlinear_classifier'])
+        self.opt_c = torch.optim.Adam([{'params': self.featurizer.parameters()}, {'params': self.classifier_c.parameters()}], 
+                                      lr=self.hparams["lr"], weight_decay=self.hparams['weight_decay'])
+        
+       
+        self.classifier_a, self.opt_a = self.create_bias_classifier(self.featurizer.n_outputs, num_classes, 
+                                                                    self.hparams['nonlinear_classifier'])
+        self.classifier_b, self.opt_b = self.create_bias_classifier(self.featurizer.n_outputs, num_classes, 
+                                                                    self.hparams['nonlinear_classifier'])
+
+        self.y_tr_dynamic = []
+        self.y_tr_true = []
+        for train_loader in train_loaders:
+            train_loader_ys = [y for _, y in train_loader.dataset.underlying_dataset]
```

Comment:
do the y's always have the same order independent of the seed?

Reply:
The order of y's is the same as what "_SplitDataset" returns samples. This dataset takes the keys and returns samples from the "underlying_dataset", I stored the same order of y's from "underlying_dataset". I changed "_SplitDataset" so that it returns the keys used to return samples and I use those indices to retrieve the "y_tr_dynamic" and "assignment" in the "update" method.

Regarding the order of "train_loaders": the order doesn't matter in this case. I am using the same order of "train_loaders" as how batches are created for the "update" method. Please look at line 216 of "train.py", where batches are created.

Reply:
great!